src/phoenix/auth.py

```diff
@@ -33,16 +61,165 @@ def validate_password_format(password: str) -> None:
+            from faker import Faker  # type: ignore[import-not-found,unused-ignore]
+
+            suggested_password = Faker().unique.password(**asdict(self))
+        except BaseException:
+            suggested_password = ""
+        if len(err_msg) > 1:
+            err_text = "Password " + ", ".join(err_msg[:-1]) + ", and " + err_msg[-1]
+        else:
+            err_text = f"Password {err_msg[0]}"
+        if suggested_password:
+            err_text = f"{err_text} (e.g. `{suggested_password}`)"
+        raise ValueError(err_text)
+
+
+"""The name of the header for token-based authentication"""
 EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+[.][^@\s]+\Z")
+"""The regular expression pattern for a valid email address."""
 NUM_ITERATIONS = 10_000
+"""The number of iterations to use for the PBKDF2 key derivation function."""
 MIN_PASSWORD_LENGTH = 4
+"""The minimum length of a password."""
+PASSWORD_REQUIREMENTS = _PasswordRequirements(MIN_PASSWORD_LENGTH)
+"""The requirements for a valid password."""
+REQUIREMENTS_FOR_PHOENIX_SECRET = _PasswordRequirements(32, True, True, True, True)
+"""The requirements for the Phoenix secret key."""
+JWT_ALGORITHM = "HS256"
+"""The algorithm to use for the JSON Web Token."""
+PHOENIX_ACCESS_TOKEN_COOKIE_NAME = "phoenix-access-token"
+PHOENIX_ACCESS_TOKEN_COOKIE_MAX_AGE = timedelta(days=31)
+
+
+class Issuer(Enum):
+    API_KEY = auto()
+    SESSION = auto()
+
+
+class ClaimStatus(Enum):
+    VALID = auto()
+    INACTIVE = auto()
+    EXPIRED = auto()
+
+
+@dataclass(frozen=True)
+class _BaseAttributes:
+    user_role: str
+
+
+@dataclass(frozen=True)
+class ApiKeyAttributes(_BaseAttributes):
+    name: Optional[str] = None
+    description: Optional[str] = None
+
+
+@dataclass(frozen=True)
+class SessionAttributes(_BaseAttributes): ...
+
+
+Attributes: TypeAlias = Union[ApiKeyAttributes, SessionAttributes]
+TokenId: TypeAlias = str
+
+
+@dataclass(frozen=True)
+class Claim:
+    token_id: Optional[TokenId] = None
+    user_id: Optional[int] = None
+    issuer: Optional[Issuer] = None
+    issued_at: Optional[datetime] = None
+    expiration_time: Optional[datetime] = None
+    audience: Optional[str] = None
+    attributes: Optional[Attributes] = None
+
+    def __post_init__(self) -> None:
+        if self.issuer is Issuer.API_KEY:
+            assert isinstance(self.attributes, ApiKeyAttributes)
+        elif self.issuer is Issuer.SESSION:
+            assert isinstance(self.attributes, SessionAttributes)
+
+    @property
+    def status(self) -> ClaimStatus:
+        if self.token_id is None:
+            return ClaimStatus.INACTIVE
+        if self.user_id is None:
+            return ClaimStatus.INACTIVE
+        if self.expiration_time and self.expiration_time < datetime.now(timezone.utc):
+            return ClaimStatus.EXPIRED
+        return ClaimStatus.VALID
+
+
+Token: TypeAlias = str
+
+
+class ApiKeyDbId(NamedTuple):
+    id_: int
+
+
+class SessionTokenDbId(NamedTuple):
+    id_: int
```

Comment:
Same as above, I'd personally avoid the term session and just make it clear that we have API keys and Access Tokens

Reply:
np, i can change. The name is just arbitrary here, although I'm not grokking to the exact subtitles between `session` and `access`: isn't `session` just a special case of `access`, i.e. time-limited?

Reply:
> A session is a group of user interactions with your website that take place within a given time frame. 

As in, there can be multiple access tokens granted during a single session.

Reply:
ah, that's the exact reverse of what i had thought. haha